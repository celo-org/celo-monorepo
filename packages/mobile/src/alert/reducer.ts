import { Actions, ActionTypes, AlertTypes, ShowAlertAction } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { ALERT_BANNER_DURATION } from 'src/config'
import {
  Actions as ExchangeActions,
  ActionTypes as ExchangeActionTypes,
} from 'src/exchange/actions'
import i18n from 'src/i18n'
import { RootState } from 'src/redux/reducers'

export enum ErrorDisplayType {
  'BANNER',
  'INLINE',
}

export type State = {
  type: 'message' | 'error'
  displayMethod: ErrorDisplayType
  message: string
  dismissAfter?: number | null
  buttonMessage?: string | null
  action?: object | null
  title?: string | null
  underlyingError?: ErrorMessages | null
} | null

const initialState = null

const errorAction = (error: ErrorMessages): ShowAlertAction => ({
  type: Actions.SHOW,
  alertType: AlertTypes.ERROR,
  displayMethod: ErrorDisplayType.BANNER,
  message: i18n.t(error, { ns: 'global' }),
  dismissAfter: ALERT_BANNER_DURATION,
  buttonMessage: null,
  title: null,
  underlyingError: error,
})

export const reducer = (
  state: State = initialState,
  action: ActionTypes | ExchangeActionTypes
): State => {
  switch (action.type) {
    case ExchangeActions.WITHDRAW_CELO_FAILED:
      action = errorAction(action.error)
    case Actions.SHOW:
      return {
        displayMethod: action.displayMethod,
        type: action.alertType,
        message: action.message,
        dismissAfter: action.dismissAfter,
        buttonMessage: action.buttonMessage,
        action: action.action,
        title: action.title,
        underlyingError: action.underlyingError,
      }
    case Actions.HIDE:
      return null
    default:
      if (state?.action === action) {
        // Hide alert when the alert action is dispatched
        return null
      }
      return state
  }
}

export const errorSelector = (state: RootState) => {
  return state.alert ? state.alert.underlyingError || null : null
}
