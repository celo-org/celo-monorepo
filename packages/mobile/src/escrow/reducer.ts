import { createSelector } from 'reselect'
import { Actions, ActionTypes, EscrowedPayment } from 'src/escrow/actions'
import { RootState } from 'src/redux/reducers'

export interface State {
  isReclaiming: boolean
  sentEscrowedPayments: EscrowedPayment[]
}

export const initialState = {
  isReclaiming: false,
  sentEscrowedPayments: [],
}

export const escrowReducer = (state: State | undefined = initialState, action: ActionTypes) => {
  switch (action.type) {
    case Actions.STORE_SENT_PAYMENTS:
      return {
        ...state,
        sentEscrowedPayments: action.sentPayments,
      }
    case Actions.RECLAIM_PAYMENT:
      return {
        ...state,
        isReclaiming: true,
      }
    case Actions.RECLAIM_PAYMENT_FAILURE:
    case Actions.RECLAIM_PAYMENT_SUCCESS:
      return {
        ...state,
        isReclaiming: false,
      }
    default:
      return state
  }
}

export const getSentEscrowPayments = (state: RootState) => state.escrow.sentEscrowedPayments

export const getReclaimableEscrowPayments = createSelector(
  getSentEscrowPayments,
  (sentPayments) => {
    const currUnixTime = Date.now() / 1000
    return sentPayments.filter((payment) => {
      const paymentExpiryTime = +payment.timestamp + +payment.expirySeconds
      return currUnixTime >= paymentExpiryTime
    })
  }
)
