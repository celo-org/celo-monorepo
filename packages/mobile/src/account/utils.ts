import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
