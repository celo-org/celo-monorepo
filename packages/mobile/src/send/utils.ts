import BigNumber from 'bignumber.js'
import { call, put, select, take } from 'redux-saga/effects'
import { showError } from 'src/alert/actions'
import { TokenTransactionType } from 'src/apollo/types'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { ALERT_BANNER_DURATION, DAILY_PAYMENT_LIMIT_CUSD } from 'src/config'
import { FeeType } from 'src/fees/actions'
import { getAddressFromPhoneNumber } from 'src/identity/contactMapping'
import { E164NumberToAddressType, SecureSendPhoneNumberMapping } from 'src/identity/reducer'
import { RecipientVerificationStatus } from 'src/identity/types'
import {
  Actions,
  FetchCurrentRateFailureAction,
  FetchCurrentRateSuccessAction,
  selectPreferredCurrency,
} from 'src/localCurrency/actions'
import { LocalCurrencyCode, LocalCurrencySymbol } from 'src/localCurrency/consts'
import { convertDollarsToLocalAmount, convertLocalAmountToDollars } from 'src/localCurrency/convert'
import { getLocalCurrencyExchangeRate } from 'src/localCurrency/selectors'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { UriData, uriDataFromUrl } from 'src/qrcode/schema'
import {
  Recipient,
  RecipientKind,
  RecipientWithContact,
  RecipientWithQrCode,
} from 'src/recipients/recipient'
import { storeLatestInRecents } from 'src/send/actions'
import { PaymentInfo } from 'src/send/reducers'
import { TransactionDataInput } from 'src/send/SendAmount'
import Logger from 'src/utils/Logger'
import { timeDeltaInHours } from 'src/utils/time'

const TAG = 'send/utils'

export interface ConfirmationInput {
  recipient: Recipient
  amount: BigNumber
  reason?: string
  recipientAddress: string | null | undefined
  type: TokenTransactionType
  firebasePendingRequestUid?: string | null
}

export const getConfirmationInput = (
  transactionData: TransactionDataInput,
  e164NumberToAddress: E164NumberToAddressType,
  secureSendPhoneNumberMapping: SecureSendPhoneNumberMapping
): ConfirmationInput => {
  const { recipient } = transactionData
  let recipientAddress: string | null | undefined

  if (recipient.kind === RecipientKind.QrCode || recipient.kind === RecipientKind.Address) {
    recipientAddress = recipient.address
  } else if (recipient.e164PhoneNumber) {
    recipientAddress = getAddressFromPhoneNumber(
      recipient.e164PhoneNumber,
      e164NumberToAddress,
      secureSendPhoneNumberMapping,
      recipient.address
    )
  }

  return { ...transactionData, recipientAddress }
}

export const getFeeType = (
  recipientVerificationStatus: RecipientVerificationStatus
): FeeType | null => {
  switch (recipientVerificationStatus) {
    case RecipientVerificationStatus.UNKNOWN:
      return null
    case RecipientVerificationStatus.UNVERIFIED:
      return FeeType.INVITE
    case RecipientVerificationStatus.VERIFIED:
      return FeeType.SEND
  }
}

// exported for tests
export function dailyAmountRemaining(now: number, recentPayments: PaymentInfo[]) {
  return DAILY_PAYMENT_LIMIT_CUSD - dailySpent(now, recentPayments)
}

function dailySpent(now: number, recentPayments: PaymentInfo[]) {
  // We are only interested in the last 24 hours
  const paymentsLast24Hours = recentPayments.filter(
    (p: PaymentInfo) => timeDeltaInHours(now, p.timestamp) < 24
  )

  const amount: number = paymentsLast24Hours.reduce((sum, p: PaymentInfo) => sum + p.amount, 0)
  return amount
}

export function isPaymentLimitReached(
  now: number,
  recentPayments: PaymentInfo[],
  initial: number
): boolean {
  const amount = dailySpent(now, recentPayments) + initial
  return amount > DAILY_PAYMENT_LIMIT_CUSD
}

export function showLimitReachedError(
  now: number,
  recentPayments: PaymentInfo[],
  localCurrencyExchangeRate: string | null | undefined,
  localCurrencySymbol: LocalCurrencySymbol | null
) {
  const dailyRemainingcUSD = dailyAmountRemaining(now, recentPayments)
  const dailyRemaining = convertDollarsToLocalAmount(dailyRemainingcUSD, localCurrencyExchangeRate)
  const dailyLimit = convertDollarsToLocalAmount(
    DAILY_PAYMENT_LIMIT_CUSD,
    localCurrencyExchangeRate
  )

  const translationParams = {
    currencySymbol: localCurrencySymbol,
    dailyRemaining,
    dailyLimit,
    dailyRemainingcUSD,
    dailyLimitcUSD: DAILY_PAYMENT_LIMIT_CUSD,
  }

  return showError(ErrorMessages.PAYMENT_LIMIT_REACHED, ALERT_BANNER_DURATION, translationParams)
}

export function* handleSendPaymentData(
  data: UriData,
  cachedRecipient?: RecipientWithContact,
  isOutgoingPaymentRequest?: true
) {
  const recipient: RecipientWithQrCode = {
    kind: RecipientKind.QrCode,
    address: data.address,
    displayId: data.e164PhoneNumber,
    displayName: data.displayName || cachedRecipient?.displayName || 'anonymous',
    e164PhoneNumber: data.e164PhoneNumber,
    phoneNumberLabel: cachedRecipient?.phoneNumberLabel,
    thumbnailPath: cachedRecipient?.thumbnailPath,
    contactId: cachedRecipient?.contactId,
  }
  yield put(storeLatestInRecents(recipient))

  if (data.currencyCode) {
    yield put(selectPreferredCurrency(data.currencyCode as LocalCurrencyCode))
    const action: FetchCurrentRateSuccessAction | FetchCurrentRateFailureAction = yield take([
      Actions.FETCH_CURRENT_RATE_SUCCESS,
      Actions.FETCH_CURRENT_RATE_FAILURE,
    ])
    if (action.type === Actions.FETCH_CURRENT_RATE_FAILURE) {
      yield put(showError(ErrorMessages.EXCHANGE_RATE_FAILED))
      Logger.warn(TAG, '@handleSendPaymentData failed to fetch current rate')
      return
    }
  }

  if (data.amount) {
    // TODO: integrate with SendConfirmation component
    const exchangeRate = yield select(getLocalCurrencyExchangeRate)
    const amount = convertLocalAmountToDollars(data.amount, exchangeRate)
    if (!amount) {
      Logger.warn(TAG, '@handleSendPaymentData null amount')
      return
    }
    const transactionData: TransactionDataInput = {
      recipient,
      amount,
      reason: data.comment,
      type: TokenTransactionType.PayPrefill,
    }
    navigate(Screens.SendConfirmation, { transactionData, isFromScan: true })
  } else {
    navigate(Screens.SendAmount, { recipient, isFromScan: true, isOutgoingPaymentRequest })
  }
}

export function* handlePaymentDeeplink(deeplink: string) {
  try {
    const paymentData = uriDataFromUrl(deeplink)
    yield call(handleSendPaymentData, paymentData)
  } catch (e) {
    Logger.warn('handlePaymentDeepLink', `deeplink ${deeplink} failed with ${e}`)
  }
}
