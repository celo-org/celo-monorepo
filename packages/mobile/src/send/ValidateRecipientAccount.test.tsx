import * as React from 'react'
import { render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import { AddressValidationType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import ValidateRecipientAccount from 'src/send/ValidateRecipientAccount'
import { createMockStore, getMockStackScreenProps } from 'test/utils'
import { mockE164NumberInvite, mockTransactionData } from 'test/values'

describe('ValidateRecipientAccount', () => {
  it('renders correctly when full validation required', () => {
    const store = createMockStore()
    const tree = render(
      <Provider store={store}>
        <ValidateRecipientAccount
          {...getMockStackScreenProps(Screens.ValidateRecipientAccount, {
            transactionData: mockTransactionData,
            addressValidationType: AddressValidationType.FULL,
          })}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when partial validation required', () => {
    const store = createMockStore()
    const tree = render(
      <Provider store={store}>
        <ValidateRecipientAccount
          {...getMockStackScreenProps(Screens.ValidateRecipientAccount, {
            transactionData: mockTransactionData,
            addressValidationType: AddressValidationType.PARTIAL,
          })}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('navigates to send confirmation when validation successful for send flow', () => {
    const store = createMockStore({
      identity: {
        secureSendPhoneNumberMapping: {
          [mockE164NumberInvite]: {
            addressValidationType: AddressValidationType.NONE,
            validationSuccessful: false,
          },
        },
      },
    })

    const props = getMockStackScreenProps(Screens.ValidateRecipientAccount, {
      transactionData: mockTransactionData,
      addressValidationType: AddressValidationType.PARTIAL,
    })

    const tree = render(
      <Provider store={store}>
        <ValidateRecipientAccount {...props} />
      </Provider>
    )

    const updatedStore = createMockStore({
      identity: {
        secureSendPhoneNumberMapping: {
          [mockE164NumberInvite]: {
            addressValidationType: AddressValidationType.NONE,
            validationSuccessful: true,
          },
        },
      },
    })

    tree.rerender(
      <Provider store={updatedStore}>
        <ValidateRecipientAccount {...props} />
      </Provider>
    )

    expect(navigate).toHaveBeenCalledWith(Screens.SendConfirmation, {
      transactionData: mockTransactionData,
      addressJustValidated: true,
    })
  })
})

it('navigates to payment request confirmation when validation successful for request flow', () => {
  const store = createMockStore({
    identity: {
      secureSendPhoneNumberMapping: {
        [mockE164NumberInvite]: {
          addressValidationType: AddressValidationType.PARTIAL,
          validationSuccessful: false,
        },
      },
    },
  })

  const props = getMockStackScreenProps(Screens.ValidateRecipientAccount, {
    transactionData: mockTransactionData,
    addressValidationType: AddressValidationType.PARTIAL,
    isOutgoingPaymentRequest: true,
  })

  const tree = render(
    <Provider store={store}>
      <ValidateRecipientAccount {...props} />
    </Provider>
  )

  const updatedStore = createMockStore({
    identity: {
      secureSendPhoneNumberMapping: {
        [mockE164NumberInvite]: {
          addressValidationType: AddressValidationType.NONE,
          validationSuccessful: true,
        },
      },
    },
  })

  tree.rerender(
    <Provider store={updatedStore}>
      <ValidateRecipientAccount {...props} />
    </Provider>
  )

  expect(navigate).toHaveBeenCalledWith(Screens.PaymentRequestConfirmation, {
    transactionData: mockTransactionData,
    addressJustValidated: true,
  })
})
