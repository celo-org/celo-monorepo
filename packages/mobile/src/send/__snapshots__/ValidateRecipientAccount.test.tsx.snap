// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidateRecipientAccount renders correctly when full validation required 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "justifyContent": "space-between",
        "padding": 20,
        "paddingTop": 40,
      }
    }
    keyboardShouldPersistTaps="always"
  >
    <View>
      <View>
        <View>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              }
            }
          >
            confirmAccountNumber.body1Full
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              }
            }
          >
            confirmAccountNumber.body2Full
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
                "lineHeight": 24,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
          >
            accountInputHeaderB
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                Object {
                  "backgroundColor": "#FFF",
                  "borderColor": "#D1D5D8",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "flex": 0,
                  "height": 50,
                  "marginVertical": 5,
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="0xf1b1d5a6e7728g309c4a025k122d71ad75a61976"
              rejectResponderTermination={true}
              shouldShowClipboard={[Function]}
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#D1D5D8",
                  "borderRadius": 3,
                  "flex": 1,
                  "fontFamily": "Hind-Regular",
                  "padding": 8,
                }
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 14,
            "lineHeight": 18,
            "marginBottom": 10,
            "marginTop": 20,
            "textAlign": "center",
            "textDecorationLine": "underline",
          }
        }
      >
        confirmAccountNumber.help
      </Text>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
        },
        null,
        null,
        undefined,
        Object {
          "backgroundColor": "#42D689",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#42D689",
          },
          Object {
            "height": 50,
          },
          Object {
            "borderWidth": 0,
          },
        ]
      }
      testID="ConfirmAccountButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
              },
              Object {
                "color": "#FFFFFF",
              },
              Object {
                "paddingLeft": 5,
                "paddingRight": 5,
              },
            ]
          }
        >
          confirmAccount.button
        </Text>
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "marginHorizontal": 10,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Bold",
              "fontSize": 18,
              "marginVertical": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.header
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.body1
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.body2
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-evenly",
              "marginTop": 25,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 14,
                    "lineHeight": 18,
                  },
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 24,
                    "paddingRight": 20,
                  },
                ]
              }
            >
              global:close
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</SafeAreaView>
`;

exports[`ValidateRecipientAccount renders correctly when partial validation required 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "justifyContent": "space-between",
        "padding": 20,
        "paddingTop": 40,
      }
    }
    keyboardShouldPersistTaps="always"
  >
    <View>
      <View>
        <View>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              }
            }
          >
            confirmAccountNumber.body1Partial
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              }
            }
          >
            confirmAccountNumber.body2Partial
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
                "lineHeight": 24,
                "marginTop": 20,
                "textAlign": "center",
              }
            }
          >
            accountInputHeaderA
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 20,
                "paddingRight": 10,
                "paddingVertical": 10,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "#FFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "borderWidth": 1,
                    "flex": 0,
                    "height": 50,
                    "marginVertical": 5,
                    "width": 50,
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                maxLength={1}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="k"
                rejectResponderTermination={true}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "flex": 1,
                    "fontFamily": "Hind-Regular",
                    "padding": 8,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "#FFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "borderWidth": 1,
                    "flex": 0,
                    "height": 50,
                    "marginVertical": 5,
                    "width": 50,
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                maxLength={1}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="0"
                rejectResponderTermination={true}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "flex": 1,
                    "fontFamily": "Hind-Regular",
                    "padding": 8,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "#FFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "borderWidth": 1,
                    "flex": 0,
                    "height": 50,
                    "marginVertical": 5,
                    "width": 50,
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                maxLength={1}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="F"
                rejectResponderTermination={true}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "flex": 1,
                    "fontFamily": "Hind-Regular",
                    "padding": 8,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "#FFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "borderWidth": 1,
                    "flex": 0,
                    "height": 50,
                    "marginVertical": 5,
                    "width": 50,
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                maxLength={1}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="4"
                rejectResponderTermination={true}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#D1D5D8",
                    "borderRadius": 3,
                    "flex": 1,
                    "fontFamily": "Hind-Regular",
                    "padding": 8,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 14,
            "lineHeight": 18,
            "marginBottom": 10,
            "marginTop": 20,
            "textAlign": "center",
            "textDecorationLine": "underline",
          }
        }
      >
        confirmAccountNumber.help
      </Text>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
        },
        null,
        null,
        undefined,
        Object {
          "backgroundColor": "#42D689",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#42D689",
          },
          Object {
            "height": 50,
          },
          Object {
            "borderWidth": 0,
          },
        ]
      }
      testID="ConfirmAccountButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
              },
              Object {
                "color": "#FFFFFF",
              },
              Object {
                "paddingLeft": 5,
                "paddingRight": 5,
              },
            ]
          }
        >
          confirmAccount.button
        </Text>
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "marginHorizontal": 10,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Bold",
              "fontSize": 18,
              "marginVertical": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.header
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.body1
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            }
          }
        >
          helpModal.body2
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-evenly",
              "marginTop": 25,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 14,
                    "lineHeight": 18,
                  },
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 24,
                    "paddingRight": 20,
                  },
                ]
              }
            >
              global:close
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</SafeAreaView>
`;
